### PR TITLE
AUTH-413: ps syncer: only sync labels if noone else is managing them

### PR DESCRIPTION
**Pull request by stlaz**
/cc @deads2k 
/cc @ibihim

### CODE REVIEW

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:306]: This doesn't do what you think it does.   You should create a new namespace

```
	corev1apply.Namespace("name").WithLabels(map[string]string{
		"one":"val",
		"two":"",
	})
```

That way your field manager only attempts to set what you want.

Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:306]: Now, separately, you've done right by not setting "force".  Now you need to allow for someone else to own these fields.  that means conflict errors are not retried.

So that apierrors.IsConflict() is handled as not an error.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:261]: I think it would be easier to debug by checking in metadata.managedFields

```yaml
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:hypershift.openshift.io/monitoring: {}
          f:openshift.io/cluster-monitoring: {}
      f:spec: {}
    manager: hypershift
    operation: Apply
    time: "2023-07-26T18:09:23Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:openshift.io/sa.scc.mcs: {}
          f:openshift.io/sa.scc.supplemental-groups: {}
          f:openshift.io/sa.scc.uid-range: {}
        f:labels:
          f:pod-security.kubernetes.io/audit: {}
          f:pod-security.kubernetes.io/audit-version: {}
          f:pod-security.kubernetes.io/warn: {}
          f:pod-security.kubernetes.io/warn-version: {}
    manager: cluster-policy-controller
    operation: Update
    time: "2023-07-26T18:09:23Z"
```

Notice how labels are subdivided by individual name (two owners in that snippet), so you can see if *your* controller set the value.

Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:261]: Isn't it better to work with the library rather than our own unstructured handling?

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:306]: The above sets "two" to an empty string. We don't want that.

From what I can see, though, if we use
```
	corev1apply.Namespace("name").WithLabels(map[string]string{
		"one":"val",
	})
```
all our managed labels but "one" will get removed.

We can use that instead, then, that should actually make the code a bit cleaner, hopefully.

Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:306]: :exclamation:  NOTE to self:
We cannot use `controllerName` as the manager because we would not be able to update fields from previous release. We must use `cluster-policy-controller` instead.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (liouk) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:340]: Duplicate `if` statement.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:556]: does this error if there are no labels owned by this particular user?

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (liouk) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:490]: This could be more readable as
```go
len(manager) > 0 && manager != "cluster-policy-controller" && manager != controllerName
```

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:521]: what is it keyed by and what's in the set?

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:556]: This only errors if it encounters an invalid structure in the path (not `map[string]interface{}`). It should never happen.

If no labels are owned, we should get `found = false`

**Comment by liouk** [comment]
/lgtm
/hold

holding for @deads2k to review

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:277]: Rather than do this, could we simply do an Apply with the `force` option set to true?

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:346]: if we are enforcing, shouldn't we check the ownership of all labels?

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:269]: Are we sure.  Aren't these used for our alerting?  Given the ability to yield to a different client, it seems like we'd want to continue setting these forever.

if we do not set these labels, then our default of "restricted" is used and this results in noise for namespaces that are not restricted because they constantly get warned and (I think), we get alerts that are not accurately indicating when a namespace will violate. 

In fact, is this leading to false positives today?

Comment (deads2k) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:269]: The concerning flow is

1. our default is restricted for audit and warn
2. the namespace is set to privileged
3. we helpfully remove the audit and warn labels
4. a privileged pod is created and fails audit and warn
5. this causes our alert about how many future violates we would have to fail.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:369]: make this a warning I think.  Given the effort, this is pretty unexpected.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:367]: reduce nesting with

```go
if is conflict
  log
  return nil
if err != nil
  return err
```

**Comment by deads2k** [comment]
worthy of an e2e test once merged.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:277]: That's not possible. The field is owned by "cluster-policy-controller + Update". You cannot set the label to an empty value (fails validation). That means that the only action you can do with it is to remove it. But I don't think there is a way to set up a `NamespaceApplyConfiguration` that would express that intention.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:269]: I think you're right and we should set all.

My original thinking back then was we do not set these fields because we're only ever going to be setting enforcement globally, and we will no longer alert at that point. I still think that that's what we should do, however, enforcement can now be enabled via feature flags while the global config stays + since the labels can now be adjusted manually separately, setting all just seems more logical.

I'll fix that.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:367]: I'd like to keep the block that contains error cases like this, unless we decide to handle more errors and introduce a switch here

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (liouk) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:340]: This probably trickled through some commit somewhere as it's still here :)

**Review by liouk — CHANGES_REQUESTED** [review]

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go]: > I'd like to keep the block that contains error cases like this, unless we decide to handle more errors and introduce a switch here

Non-standard with both golang and kube.  Sort of like using `== ""` versus `len(foo) == 0`.

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (deads2k) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:277]: huh

Comment (deads2k) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:277]: > That's not possible. The field is owned by "cluster-policy-controller + Update". You cannot set the label to an empty value (fails validation). That means that the only action you can do with it is to remove it. But I don't think there is a way to set up a `NamespaceApplyConfiguration` that would express that intention

Can't you set it to the current value with a force?

Comment (stlaz) [reply pkg/psalabelsyncer/podsecurity_label_sync_controller.go:277]: If you set it to the current value, the "force" does not seem to matter at all. You will always get double ownership, which will still result in a conflict when you're trying to change it. Some fun corner cases I tested:

1. `cluster-policy-controller` + apply to the same value
    a. the labels is now managed by `cluster-policy-controller + update` and `cluster-policy-controller + apply`
    b. you get the applyconfig and remove the ownership in an apply as the cluster-policy-controller
    c. the field is now again owned only by `cluster-policy-controller + update`
2. `this-controller` + apply  the same value
   a. the label is now managed by `cluster-policy-controller + update` and `this-controller + apply`
   b. try to change the label value as this-controller - you get a conflict

File: pkg/psalabelsyncer/podsecurity_label_sync_controller.go
Comment (liouk) [inline pkg/psalabelsyncer/podsecurity_label_sync_controller.go:344]: @stlaz just realized that this comment got somehow truncated, just FYI

**Pull request merged**
